File: src/vm.c
/*
 * vm.c — Stack-based virtual machine.
 *
 * Recursion support:
 *   OP_CALL pushes a call frame (return address + old frame pointer).
 *   OP_ENTER allocates space for local variables.
 *   OP_RET restores the previous frame and pushes the return value.
 *   Each recursive call gets its own frame with isolated variables.
 */

#include "vm.h"

typedef struct {
    int return_ip;
    int fp;
} CallFrame;

int vm_execute(const Program *prog) {
    int        stack[MAX_STACK];
    CallFrame  frames[MAX_FRAMES];
    int        sp = 0;           /* stack pointer          */
    int        fp = 0;           /* frame pointer          */
    int        ip = 0;           /* instruction pointer    */
    int        fc = 0;           /* frame count            */
    const int *code = prog->code;

    for (;;) {
        if (ip < 0 || ip >= prog->count)
            error_exit(0, "VM: instruction pointer out of bounds (%d)", ip);

        int op = code[ip++];

        switch (op) {

        /* ── Stack ────────────────────────────────────── */
        case OP_CONST:
            stack[sp++] = code[ip++];
            break;

        case OP_POP:
            sp--;
            break;

        /* ── Arithmetic ──────────────────────────────── */
        case OP_ADD: { int b=stack[--sp]; int a=stack[--sp]; stack[sp++]=a+b; break; }
        case OP_SUB: { int b=stack[--sp]; int a=stack[--sp]; stack[sp++]=a-b; break; }
        case OP_MUL: { int b=stack[--sp]; int a=stack[--sp]; stack[sp++]=a*b; break; }
        case OP_DIV: {
            int b=stack[--sp]; int a=stack[--sp];
            if (b == 0) error_exit(0, "VM: division by zero");
            stack[sp++]=a/b; break;
        }
        case OP_MOD: {
            int b=stack[--sp]; int a=stack[--sp];
            if (b == 0) error_exit(0, "VM: modulo by zero");
            stack[sp++]=a%b; break;
        }
        case OP_NEG: stack[sp-1] = -stack[sp-1]; break;

        /* ── Comparison / Logic ──────────────────────── */
        case OP_EQ:  { int b=stack[--sp]; int a=stack[--sp]; stack[sp++]=(a==b); break; }
        case OP_NEQ: { int b=stack[--sp]; int a=stack[--sp]; stack[sp++]=(a!=b); break; }
        case OP_LT:  { int b=stack[--sp]; int a=stack[--sp]; stack[sp++]=(a<b);  break; }
        case OP_GT:  { int b=stack[--sp]; int a=stack[--sp]; stack[sp++]=(a>b);  break; }
        case OP_LTE: { int b=stack[--sp]; int a=stack[--sp]; stack[sp++]=(a<=b); break; }
        case OP_GTE: { int b=stack[--sp]; int a=stack[--sp]; stack[sp++]=(a>=b); break; }
        case OP_AND: { int b=stack[--sp]; int a=stack[--sp]; stack[sp++]=(a&&b)?1:0; break; }
        case OP_OR:  { int b=stack[--sp]; int a=stack[--sp]; stack[sp++]=(a||b)?1:0; break; }
        case OP_NOT: stack[sp-1] = !stack[sp-1]; break;

        /* ── Variables ───────────────────────────────── */
        case OP_LOAD: {
            int slot = code[ip++];
            stack[sp++] = stack[fp + slot];
            break;
        }
        case OP_STORE: {
            int slot = code[ip++];
            stack[fp + slot] = stack[--sp];
            break;
        }

        /* ── Control flow ────────────────────────────── */
        case OP_JMP:
            ip = code[ip];
            break;

        case OP_JZ: {
            int addr = code[ip++];
            if (stack[--sp] == 0) ip = addr;
            break;
        }

        /* ── Function calls (key to recursion!) ──────── */
        case OP_CALL: {
            int addr   = code[ip++];
            int nargs  = code[ip++];

            /* Save current frame */
            if (fc >= MAX_FRAMES)
                error_exit(0, "VM: call stack overflow (too much recursion?)");
            frames[fc].return_ip = ip;
            frames[fc].fp        = fp;
            fc++;

            /* New frame: fp points to first argument */
            fp = sp - nargs;
            ip = addr;
            break;
        }

        case OP_RET: {
            int retval = stack[--sp];

            /* Discard frame (args + locals) */
            sp = fp;

            /* Restore caller frame */
            fc--;
            ip = frames[fc].return_ip;
            fp = frames[fc].fp;

            /* Push return value for caller */
            stack[sp++] = retval;
            break;
        }

        case OP_ENTER: {
            int nlocals = code[ip++];
            /* Reserve space for local variables (init to 0) */
            for (int i = 0; i < nlocals; i++)
                stack[sp++] = 0;
            break;
        }

        /* ── Built-ins ───────────────────────────────── */
        case OP_PRINT: {
            int val = stack[--sp];
            printf("%d\n", val);
            fflush(stdout);
            stack[sp++] = 0;   /* push dummy return value */
            break;
        }

        case OP_SCAN: {
            int val = 0;
            printf("> ");
            fflush(stdout);
            if (scanf("%d", &val) != 1)
                error_exit(0, "VM: failed to read integer from input");
            stack[sp++] = val;
            break;
        }

        case OP_HALT:
            /* Return value of main is on the stack */
            return (sp > 0) ? stack[sp - 1] : 0;

        default:
            error_exit(0, "VM: unknown opcode %d at position %d", op, ip - 1);
        }

        /* Stack bounds check */
        if (sp < 0)
            error_exit(0, "VM: stack underflow");
        if (sp >= MAX_STACK)
            error_exit(0, "VM: stack overflow");
    }
}

File: desktop/compiler/vm.js
/**
 * vm.js — Direct port of src/vm.c
 * Stack-based virtual machine with instrumentation for visualization.
 */

const { OpCode, opName } = require('./compiler');
const { CompileError } = require('./lexer');

const MAX_STACK = 4096;
const MAX_FRAMES = 256;

class VMError extends Error {
    constructor(message) {
        super(`VM: ${message}`);
        this.name = 'VMError';
    }
}

/**
 * Execute a compiled program.
 *
 * Options:
 *   onOutput(value)          — called when OP_PRINT fires
 *   onInput()                — called when OP_SCAN fires, must return a number
 *   onCall(funcName, args, depth)   — called when OP_CALL fires
 *   onReturn(funcName, retval, depth) — called when OP_RET fires
 *   funcTable                — array of {name, address, nParams} from compiler
 */
function vmExecute(prog, options = {}) {
    const {
        onOutput = (v) => {},
        onInput = () => 0,
        onCall = () => {},
        onReturn = () => {},
        funcTable = [],
    } = options;

    const stack = new Array(MAX_STACK).fill(0);
    const frames = [];
    let sp = 0;
    let fp = 0;
    let ip = 0;
    let fc = 0;
    const code = prog.code;

    // Build address-to-function-name map
    const addrToFunc = {};
    for (const f of funcTable) {
        addrToFunc[f.address] = f.name;
    }

    // Track call stack for visualization
    const callStackHistory = [];
    const currentCallStack = [];
    let maxDepth = 0;

    const outputs = [];
    let stepCount = 0;
    const MAX_STEPS = 1000000; // safety limit

    for (;;) {
        if (ip < 0 || ip >= prog.count) {
            throw new VMError(`instruction pointer out of bounds (${ip})`);
        }

        if (++stepCount > MAX_STEPS) {
            throw new VMError('execution limit reached (possible infinite loop)');
        }

        const op = code[ip++];

        switch (op) {
            case OpCode.OP_CONST:
                stack[sp++] = code[ip++];
                break;

            case OpCode.OP_POP:
                sp--;
                break;

            case OpCode.OP_ADD: { const b=stack[--sp]; const a=stack[--sp]; stack[sp++]=a+b; break; }
            case OpCode.OP_SUB: { const b=stack[--sp]; const a=stack[--sp]; stack[sp++]=a-b; break; }
            case OpCode.OP_MUL: { const b=stack[--sp]; const a=stack[--sp]; stack[sp++]=a*b; break; }
            case OpCode.OP_DIV: {
                const b=stack[--sp]; const a=stack[--sp];
                if (b === 0) throw new VMError('division by zero');
                stack[sp++] = Math.trunc(a/b); break;
            }
            case OpCode.OP_MOD: {
                const b=stack[--sp]; const a=stack[--sp];
                if (b === 0) throw new VMError('modulo by zero');
                stack[sp++] = a%b; break;
            }
            case OpCode.OP_NEG: stack[sp-1] = -stack[sp-1]; break;

            case OpCode.OP_EQ:  { const b=stack[--sp]; const a=stack[--sp]; stack[sp++]=(a===b)?1:0; break; }
            case OpCode.OP_NEQ: { const b=stack[--sp]; const a=stack[--sp]; stack[sp++]=(a!==b)?1:0; break; }
            case OpCode.OP_LT:  { const b=stack[--sp]; const a=stack[--sp]; stack[sp++]=(a<b)?1:0;  break; }
            case OpCode.OP_GT:  { const b=stack[--sp]; const a=stack[--sp]; stack[sp++]=(a>b)?1:0;  break; }
            case OpCode.OP_LTE: { const b=stack[--sp]; const a=stack[--sp]; stack[sp++]=(a<=b)?1:0; break; }
            case OpCode.OP_GTE: { const b=stack[--sp]; const a=stack[--sp]; stack[sp++]=(a>=b)?1:0; break; }
            case OpCode.OP_AND: { const b=stack[--sp]; const a=stack[--sp]; stack[sp++]=(a&&b)?1:0; break; }
            case OpCode.OP_OR:  { const b=stack[--sp]; const a=stack[--sp]; stack[sp++]=(a||b)?1:0; break; }
            case OpCode.OP_NOT: stack[sp-1] = (!stack[sp-1])?1:0; break;

            case OpCode.OP_LOAD: {
                const slot = code[ip++];
                stack[sp++] = stack[fp + slot];
                break;
            }
            case OpCode.OP_STORE: {
                const slot = code[ip++];
                stack[fp + slot] = stack[--sp];
                break;
            }

            case OpCode.OP_JMP:
                ip = code[ip];
                break;

            case OpCode.OP_JZ: {
                const addr = code[ip++];
                if (stack[--sp] === 0) ip = addr;
                break;
            }

            case OpCode.OP_CALL: {
                const addr = code[ip++];
                const nargs = code[ip++];

                if (fc >= MAX_FRAMES)
                    throw new VMError('call stack overflow (too much recursion?)');

                frames[fc] = { return_ip: ip, fp: fp };
                fc++;

                // Collect args for visualization
                const args = [];
                for (let i = 0; i < nargs; i++) {
                    args.push(stack[sp - nargs + i]);
                }

                fp = sp - nargs;
                ip = addr;

                const funcName = addrToFunc[addr] || `func@${addr}`;
                currentCallStack.push({ name: funcName, args: [...args], depth: fc });
                maxDepth = Math.max(maxDepth, fc);

                callStackHistory.push({
                    action: 'call',
                    name: funcName,
                    args: [...args],
                    depth: fc,
                    stack: currentCallStack.map(f => ({ ...f })),
                });

                onCall(funcName, args, fc);
                break;
            }

            case OpCode.OP_RET: {
                const retval = stack[--sp];
                sp = fp;
                fc--;
                ip = frames[fc].return_ip;
                fp = frames[fc].fp;
                stack[sp++] = retval;

                const frame = currentCallStack.pop();
                const funcName = frame ? frame.name : '?';

                callStackHistory.push({
                    action: 'return',
                    name: funcName,
                    returnValue: retval,
                    depth: fc,
                    stack: currentCallStack.map(f => ({ ...f })),
                });

                onReturn(funcName, retval, fc);
                break;
            }

            case OpCode.OP_ENTER: {
                const nlocals = code[ip++];
                for (let i = 0; i < nlocals; i++) {
                    stack[sp++] = 0;
                }
                break;
            }

            case OpCode.OP_PRINT: {
                const val = stack[--sp];
                outputs.push(val);
                onOutput(val);
                stack[sp++] = 0;
                break;
            }

            case OpCode.OP_SCAN: {
                const val = onInput();
                if (typeof val !== 'number' || isNaN(val)) {
                    throw new VMError('scan() requires a valid integer input');
                }
                stack[sp++] = Math.trunc(val);
                break;
            }

            case OpCode.OP_HALT:
                return {
                    exitCode: (sp > 0) ? stack[sp - 1] : 0,
                    outputs,
                    callStackHistory,
                    maxRecursionDepth: maxDepth,
                    stepCount,
                };

            default:
                throw new VMError(`unknown opcode ${op} at position ${ip - 1}`);
        }

        if (sp < 0) throw new VMError('stack underflow');
        if (sp >= MAX_STACK) throw new VMError('stack overflow');
    }
}

if (typeof module !== 'undefined') {
    module.exports = { vmExecute, VMError };
}
